src/lib.rs

```diff
@@ -275,14 +282,88 @@ fn parse_unary_operator(input: ParseInput) -> ParseResult<UnaryOperator> {
     spanned(op).parse(input)
 }
 
-/// Parses unary and atomic expressions. E.g., `-u1:1`, `u1:1`
+/// Parses a binary operator. E.g. `|`, `&`, etc.
+fn parse_binary_operator(input: ParseInput) -> ParseResult<BinaryOperator> {
+    let op = alt((
+        // These two must match before | and &
+        value(RawBinaryOperator::BooleanOr, tag("||")),
+        value(RawBinaryOperator::BooleanAnd, tag("&&")),
+        // now match | and &
+        value(RawBinaryOperator::BitwiseOr, tag("|")),
+        value(RawBinaryOperator::BitwiseAnd, tag("&")),
+        value(RawBinaryOperator::BitwiseXor, tag("^")),
+        // concatenate must match before +
+        value(RawBinaryOperator::Concatenate, tag("++")),
+        // the order of these does not matter
+        // arithmetic
+        value(RawBinaryOperator::Add, tag("+")),
+        value(RawBinaryOperator::Subtract, tag("-")),
+        value(RawBinaryOperator::Multiply, tag("*")),
+        // shift
+        value(RawBinaryOperator::ShiftRight, tag(">>")),
+        value(RawBinaryOperator::ShiftLeft, tag("<<")),
+        // compare
+        value(RawBinaryOperator::Equal, tag("==")),
+        value(RawBinaryOperator::NotEqual, tag("!=")),
+        value(RawBinaryOperator::GreaterOrEqual, tag(">=")),
+        value(RawBinaryOperator::Greater, tag(">")),
+        value(RawBinaryOperator::LessOrEqual, tag("<=")),
+        value(RawBinaryOperator::Less, tag("<")),
+    ));
+    spanned(op).parse(input)
+}
+
+/// Parses a let expression. E.g. `let x : u32 = a + u32:1; x`
+///
+/// Note the trailing expression `x`. It is part of the let expression. It is common to have an
+/// expression after the let (after all, what's the point of declaring a variable binding if
+/// you're never going to use the variable?), but not required. So the trailing expression is
+/// optional.
+fn parse_let_expression(
+    input: ParseInput,
+) -> ParseResult<(NonEmpty<LetBinding>, Option<Expression>)> {
+    fn parse_let_binding(input: ParseInput) -> ParseResult<RawLetBinding> {
+        let var_decl = delimited(
+            // let must be followed by at least 1 whitespace
+            tuple((tag_ws("let"), whitespace_exactly1)),
+            parse_variable_declaration,
+            tag_ws("="),
+        );
+        let bound_expr = terminated(parse_expression(None), tag_ws(";"));
+        tuple((var_decl, bound_expr))
+            .map(|(variable_declaration, value)| RawLetBinding {
+                variable_declaration,
+                value: Box::new(value),
+            })
+            .parse(input)
+    }
+
+    // We avoid a recursive parsing implementation of nested let expressions to avoid stack
+    // overflows when fuzzing.
```

Comment:
Would[n't] it be simpler for a fuzzer to just have "max/current depth" elements in the context in the generator?

Reply:
Rust can still be hard to read for me, so consider the source :)

Reply:
Personally, I would rather our parser be as robust as possible. You never know when a user is going to do something crazy. E.g. maybe the user has a source code generator, so they're making heavily nested let expressions. We should parse everything that's valid.

Besides, it was extremely simple to make a non-recursive implementation by using the `many1` combinator here. In fact, using `many1` was easier than making a recursive implementation.